crates/ide/src/goto_definition.rs

```diff
@@ -44,6 +39,12 @@ pub(crate) fn goto_definition(
                 let nav = def.try_to_nav(sema.db)?;
                 vec![nav]
             },
+            ast::SelfParam(self_param) => {
+                let ty = sema.type_of_self(&self_param)?;
+                let adt_def = ty.autoderef(db).filter_map(|ty| ty.as_adt()).last()?;
```

Comment:
This sounds like it won't work for the `impl u8` etc. in libcore, but that's alright